Sephora-M/assignment2/BinaryTree.java

```diff
@@ -0,0 +1,285 @@
+package assignment2;
+
+/**
+ * This class implements a Binary Tree. Each node in the tree is an object BinaryTreeNode<T> that is represented by a key 
+ * of generic type T and points to its children and to its parent. It is assumed that no two nodes have the same key. 
+ * 
+ * @author Sephora
+ *
+ * @param <T>
+ */
+import java.util.HashSet;
+import java.util.NoSuchElementException;
+import java.util.Stack;
+import java.util.concurrent.ThreadLocalRandom;
+
+public class BinaryTree<T> {
+	private BinaryTreeNode<T> root;
+	private HashSet<T> keys = new HashSet<T>();
+	private T openDelimiter;
+	private T closeDelimiter;
+	
+	/**
+	 * This constructor initialises a tree with a single node, the root node.
+	 * @param root
+	 */
+	public BinaryTree(T rootKey) {
+		BinaryTreeNode<T> root = new BinaryTreeNode<T>(rootKey);
+		this.root = root;
+		keys.add(rootKey);
+	}
+	
+	/**
+	 * This constructor initialises a tree with from a list of node keys given pre-
+	 .
+	 * It uses user defined delimiters to specify the tree hierarchy. For examples, if the 
+	 * delimiters are "(" and ")", the simple tree having "a" as a root and "b" and "c" as children
+	 * is described by the list ["a", "(", "(","b", ")", "(", "c", ")",")"]   
+	 * @param preOrderKeys	T[], the pre-order list of node keys
+	 * @param openDelimiter the open delimiter, a type T object
+	 * @param closeDelimiter the close delimiter, a type T object
+	 */
+	public BinaryTree(T[] preOrderKeys, T openDelimiter, T closeDelimiter) {
```

Comment:
optional: In your notation, how would you represent a tree like this?
```
        a
      /   \
    b      c
     \    /   
      d  e
           \
            f
```

Reply:
Well in fact because I don't take into account the ordering of the children I would see it more like that:
        a
      /   \
    b      c
    |       |
    d      e
            |            
            f

Which would be represented as [a, (, b, (, d, ), ), (, c, (, e, (, f, ), ), )], where I've omitted the quotes for readability. 

Now if the order of the children is important, I would represent it as [a, (, b, (, null, ), (, d, ), ), (, c, (, e, (, null, ), (, f, ), ), )]. I think my implementation can support adding a child with null key (but I'd have to take care of the keysSet, make sure we can add multiple null keys..)

Reply:
@imorlowska Did you want to comment on this before I delete this branch? :)